common/src/main/java/com/wynntils/wc/utils/lootrun/LootrunUtils.java

```diff
@@ -0,0 +1,814 @@
+                return notes.remove(i);
+            }
+        }
+        return null;
+    }
+
+    public static List<Pair<Vec3, Component>> getCurrentNotes() {
+        LootrunUncompiled activeLootrun = getActiveLootrun();
+
+        if (activeLootrun != null) return activeLootrun.notes();
+
+        return new ArrayList<>();
+    }
+
+    public static void setLastChestIfRecording(BlockPos pos) {
+        if (state != LootrunState.RECORDING) {
+            return;
+        }
+
+        recordingInformation.setLastChest(pos);
+    }
+
+    public static void addChestIfRecording() {
+        if (state != LootrunState.RECORDING) {
+            return;
+        }
+
+        if (recordingInformation.lastChest == null) {
+            return;
+        }
+
+        recording.chests().add(recordingInformation.getLastChest());
+        recordingInformation.setDirty(true);
+        recordingInformation.setLastChest(null);
+    }
+
+    public static void recordMovementIfRecording() {
+        if (state != LootrunState.RECORDING) {
+            return;
+        }
+
+        LocalPlayer player = McUtils.mc().player;
+        if (player != null) {
+            Entity root = player.getRootVehicle();
+            Vec3 pos = root.position();
+            if (recordingInformation.getLastLocation() == null
+                    || pos.distanceToSqr(recordingInformation.getLastLocation()) >= 4d) {
+                recording.points().add(pos);
+                recordingInformation.setLastLocation(pos);
+                recordingInformation.setDirty(true);
+            }
+
+            if (recordingInformation.isDirty()) {
+                recordingCompiled = compile(recording);
+                recordingInformation.setDirty(false);
+            }
+        }
+    }
+
+    public static Vec3 getStartingPoint() {
+        LootrunUncompiled activeLootrun = getActiveLootrun();
+
+        if (activeLootrun == null) {
+            return null;
+        }
+
+        if (activeLootrun.points == null || activeLootrun.points.size() == 0) return null;
+
+        return activeLootrun.points.get(0);
+    }
+
+    public enum LootrunSaveResult {
+        SAVED,
+        ERROR_SAVING,
+        ERROR_ALREADY_EXISTS
+    }
+
+    public enum LootrunUndoResult {
+        SUCCESSFUL,
+        ERROR_STAND_NEAR_POINT,
+        ERROR_NOT_FAR_ENOUGH
+    }
+
+    public enum LootrunState {
+        RECORDING, // Lootrun is being recorded
+        LOADED, // Lootrun is loaded and displayed
+        DISABLED // No lootrun paths are rendered or loaded
+    }
+
+    public record LootrunInstance(
+            Long2ObjectMap<List<List<ColoredPoint>>> points,
+            Long2ObjectMap<Set<BlockPos>> chests,
+            Long2ObjectMap<List<Pair<Vec3, Component>>> notes) {}
+
+    public record ColoredPoint(Vec3 vec3, int color) {}
+
+    public static class RecordingInformation {
+        private Vec3 lastLocation;
+        private BlockPos lastChest;
```

Comment:
where is `lastLocation` and `lastChest` used?

Reply:
They are used in recording. `LootrunUtils#729 and 727`